### DISCUSSION TITLE
Getting custom box drawing characters to line up perfectly

### DISCUSSION THREAD

**Discussion by rbong** [kovidgoyal/kitty#7680] [Q&A] [answered]
I am trying to essentially create custom "box drawing" characters for a terminal-based branch viewer.

The issue I'm running into is that characters don't line up perfectly. Essentially, I can never get the box drawing characters to "touch". There is always a small amount of space between them, either vertically or horizontally or both.

For instance, this character is the closest I've gotten to filling one whole character space, but there are slight gaps between the characters horizontally and vertically:

![2024-07-29-175106_104x136_scrot](https://github.com/user-attachments/assets/473f717e-7e95-4744-bf40-ddd19d15ff8c)

Of course, this is only one font size. On other font sizes, they line up vertically, but not horizontally, or vise-versa:

![2024-07-29-175247_202x248_scrot](https://github.com/user-attachments/assets/9889721f-7604-4d29-bc10-0e07a6b1d4f9)

I assume this is because of rounding true font sizes to pixel sizes. But it's not clear to me:

* What the ratio of font width to height is or what factors influence this, either while creating the font or configuring fonts in Kitty.
* What influences the effective outer bounds of glyphs, since you seem to need to go out of the glyph bounds in FontForge to get them close to lining up in Kitty.
* If it's possible to create box drawing glyphs that line up at most or all font sizes.

I am aware that most terminals generate their own box drawing characters, Kitty included, because this is a hard problem, but that's why I'm asking for help. At this point, until I know the answers to the above questions, I can't even get close.

I am also aware this is not a Kitty-specific problem. But Kitty has the best font rendering I have tried. My goal right now is just to get it right in one font size in one terminal. If there is a more "font creating and rendering" specific forum I could ask this same question, I'm happy to move it there.

Here's the font in case it helps:

[font.zip](https://github.com/user-attachments/files/16419602/font.zip)

The box character is "", or unicode character 0xf640

My config is as follows:

```
include ~/.config/kitty/colors/gruvbox.conf

disable_ligatures always

font_family 0xProto Nerd Font
font_size   12.0

placement_strategy top-left
```

For now I will keep researcing, testing, and digging into the the source code to solve this issue.

**Comment by kovidgoyal** [comment, marked as answer]
IME, you cannot. Fonts rendering simply isnt designed for this, there is all sorts of fitting, hinting, etc. that goes on that means trying to line up complex shapes correctly at all sizes, hinting settings is a lost cause. 

This is why kitty generates box drawing character programmatically. It was among the first terminals to do so, nowadays all modern terminals follow it. 

Your best bet would be simply to get your characters included in something like NERD fonts and then hopefully terminals will eventually pick them up as additional box drawing characters. kitty for instance draws a number of characters from powerline and nerd font programmatically. You can try contributing code to box_drawing.py to draw your characters, since it will be python code it will be easy to follow for other terminal emulators if and when they decide to add them.

**Reply by rbong** [reply]
Thank you for the reply. I thought this might be the case, but you saved me a lot of time confirming this.

**Reply by rbong** [reply]
Sorry, one more thing. I'm not sure if the path you've outlined for getting these symbols into terminals will work.

Getting this merged into Nerd Fonts is a bit of a chicken-and-egg problem. Nerd Fonts aggregates *popular* font icons, and if what you're saying is true, it's not possible to create a program which can effectively use branch drawing icons to become popular in the first place, since it's impossible to get them to line up reliably.

I'm guessing if I kick this over to Nerd Fonts, they will probably shut this down. So I'd like to ask; if I put together an MR that generates branch drawing icons for a terminal-based Git branch viewer, would you consider accepting it?

These are the symbols so far:

![flog-symbols](https://github.com/user-attachments/assets/10575e66-130f-439c-afc5-cab517472e14)

I included symbols that will work with a git graph oriented in any direction, not just what I need.

Just to show there is potential interest; this is for my established vim/nvim branch viewer [vim-flog](https://github.com/rbong/vim-flog). There is also [someome working on an Nvim branch viewer who wants to use symbols exactly like this](https://old.reddit.com/r/neovim/comments/1ec118n/git_graph_teaser/), and it is fairly well anticipated.

**Reply by cavanaug** [reply]
Did this ever get merged into a nerd font?

**Reply by rbong** [reply]
No, because there's so many issues getting the symbols to line up, it has limited usefulness if it gets merged into a font. There are almost guaranteed rendering issues unless if font rendering is perfect, so it's best if your terminal emulator supports the symbols directly.

If you want to try the font route, you can try [flog symbols](https://github.com/rbong/flog-symbols/).

If you'd like to try a different terminal than Kitty, I believe these characters are supported by Countour, Ghostty, and WezTerm.

**Comment by kovidgoyal** [comment]
Sure, just be careful when assigning the private use unicode code
points. Don't trample on ones used by other popular iconsets.

A good start is to avoid everything used by NERD fonts, but also look at
Fira Code, powerline, etc.
